botfront/imports/ui/components/templates/templates-list/BotResponseEditor.jsx

```diff
@@ -0,0 +1,322 @@
+import {
+    Segment, Menu, MenuItem, Modal,
+} from 'semantic-ui-react';
+// connections
+import { CREATE_BOT_RESPONSE, UPDATE_BOT_RESPONSE } from '../mutations';
+import { RESPONSES_MODIFIED } from './subscriptions';
+import { GET_BOT_RESPONSE } from '../queries';
+// components
+import { ProjectContext } from '../../../layouts/context';
+import SequenceEditor from './SequenceEditor';
+import MetadataForm from '../MetadataForm';
+import ResponseNameInput from '../common/ResponseNameInput';
+// utils
+import { createResponseFromTemplate, checkResponseEmpty } from '../botResponse.utils';
+import { clearTypenameField } from '../../../../lib/utils';
+
+
+/*
+Bot response Editor requireds one of: botResponse, name, or isNew.
+botResponse is a full bot response object passed from its parent
+name is the response key which bot response editor will use to fetch the response
+isNew will create a new response that is saved on modal close
+*/
+
+const BotResponseEditor = (props) => {
+    const {
+        botResponse,
+        open,
+        trigger,
+        closeModal,
+        renameable,
+        isNew,
+        language,
+        projectId,
+        refreshBotResponse,
+        name,
+    } = props;
+
+    const { upsertResponse } = useContext(ProjectContext); // using the upsert function from the project context ensures the visual story is updated
+    const [createBotResponse] = useMutation(CREATE_BOT_RESPONSE);
+    const [updateBotResponse] = useMutation(UPDATE_BOT_RESPONSE);
+    
+    const [newBotResponse, setNewBotResponse] = useState();
+    const [activeTab, setActiveTab] = useState(0);
+    const [responseKey, setResponseKey] = useState(botResponse.key);
+    const [renameError, setRenameError] = useState();
+
+    const insertResponse = (newResponse, callback) => {
+        createBotResponse({
+            variables: {
+                projectId,
+                response: clearTypenameField(newResponse)
+            }
+        }).then(
+            (result) => { callback(undefined, result); },
+            (error) => { callback(error); },
+        );
+    };
+
+    const updateResponse = (updatedResponse, callback) => {
+        updateBotResponse({
+            variables: {
+                projectId, _id: updatedResponse._id, response: clearTypenameField(updatedResponse),
+            },
+        }).then(
+            (result) => {
+                callback(undefined, result);
+            },
+            (error) => {
+                callback(error);
+            },
+        );
+    };
+
+    const handleChangeMetadata = (updatedMetadata) => {
+        if (isNew) {
+            setNewBotResponse(
+                { ...(newBotResponse || botResponse), metadata: updatedMetadata });
+            return;
+        }
+        updateResponse({ ...botResponse, metadata: updatedMetadata }, () => {});
+    };
+
+    const handleChangeKey = async () => {
+        if (!responseKey.match(/^utter_/)) {
```

Comment:
Shouldn't that be enforced at the schema level?

Reply:
this is now enforced at the schema level